core/market/decentralized/src/matcher/matcher.rs

```diff
@@ -4,21 +4,60 @@ use thiserror::Error;
 use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
 
 use ya_client::model::market::{Demand, Offer, Proposal};
+use ya_client::model::ErrorMessage;
 use ya_persistence::executor::DbExecutor;
 use ya_persistence::executor::Error as DbError;
 
-use crate::protocol::{Discovery, DiscoveryBuilder, DiscoveryFactory, DiscoveryInitError};
+use crate::db::dao::*;
+use crate::db::models::Demand as ModelDemand;
+use crate::db::models::Offer as ModelOffer;
+use crate::db::*;
+use crate::migrations;
+use crate::protocol::{
+    Discovery, DiscoveryBuilder, DiscoveryError, DiscoveryFactory, DiscoveryInitError,
+};
 use crate::protocol::{OfferReceived, RetrieveOffers};
 
 #[derive(Error, Debug)]
-pub enum MatcherError {}
+pub enum DemandError {
+    #[error("Failed to insert Demand. Error: {}.", .0)]
+    InsertDemandFailure(#[from] DbError),
+    #[error("Failed to remove Demand [{}]. Error: {}.", .1, .0)]
+    RemoveDemandFailure(DbError, String),
+    #[error("Demand [{}] doesn't exist.", .0)]
+    DemandNotExists(String),
+}
+
+#[derive(Error, Debug)]
+pub enum OfferError {
+    #[error("Failed to insert Offer. Error: {}.", .0)]
+    InsertOfferFailure(#[from] DbError),
+    #[error("Failed to remove Offer [{}]. Error: {}.", .1, .0)]
+    RemoveOfferFailure(DbError, String),
+    #[error("Offer [{}] doesn't exist.", .0)]
+    OfferNotExists(String),
+    #[error("Failed to broadcast offer [{}]. Error: {}.", .0, .1)]
+    BroadcastOfferFailure(DiscoveryError, String),
+}
+
+#[derive(Error, Debug)]
+pub enum MatcherError {
+    #[error(transparent)]
+    DemandError(#[from] DemandError),
+    #[error(transparent)]
+    OfferError(#[from] OfferError),
+    #[error("Internal error: {}.", .0)]
+    InternalError(String),
+}
 
 #[derive(Error, Debug)]
 pub enum MatcherInitError {
     #[error("Failed to initialize Discovery interface. Error: {}.", .0)]
     DiscoveryError(#[from] DiscoveryInitError),
     #[error("Failed to initialize database. Error: {}.", .0)]
     DatabaseError(#[from] DbError),
+    #[error("Failed to migrate market database. Error: {}.", .0)]
+    MigrationError(#[from] anyhow::Error),
 }
 
 /// Receivers for events, that can be emitted from Matcher.
```

Comment:
missed that before, but I'd rather use generics instead of `dyn` in Matcher struct: [a few lines below](https://github.com/golemfactory/yagna/pull/269/files#diff-c373b03f716ab716b09699faa1c50599R70)

Reply:
Which one _dyn_? This link doesn't point to anything :(

Reply:
Ok, probably you meant Arc<dyn Discovery>.
It will disappear in https://github.com/golemfactory/yagna/pull/294, because thsi abstraction won't be necessary anymore.

But meanwhile, when we are speaking about this, I totally disagree, that it would be better.
If Matcher was generic, all structs that reference it, would have to be generic too. In general case it means generics explosion in top level struct (Market in this case). Maybe in current implementation it wouldn't happen, but you were talking here about general rules here, as I understand.
And what are actually disadvantages of `dyn` that are important here?